### PR TITLE
No team cycles

### PR DESCRIPTION
**Pull request by cbart**
This change introduces a check on `UpdateTeam` to not introduce a circular parent relationship.

The check effectively finds all teams that _are not descendants_ (so not in the subtree of) modified team, and looks whether the parent is there. It is only safe to set the parent team ID to a team outside of modified team's own descendants.

## Test plan

*   Testing database stores with Postgres.
*   Testing GraphQL API with fakes.

### CODE REVIEW

File: internal/database/teams.go [hunk 1]
```diff
@@ -51,8 +53,22 @@ func (opts ListTeamsOpts) SQL() (where, joins []*sqlf.Query) {
 		joins = append(joins, sqlf.Sprintf("JOIN team_members ON team_members.team_id = teams.id"))
 		where = append(where, sqlf.Sprintf("team_members.user_id = %s", opts.ForUserMember))
 	}
+	if opts.ExceptAncestorID != 0 {
+		joins = append(joins, sqlf.Sprintf("LEFT JOIN descendants ON descendants.id = teams.id"))
+		where = append(where, sqlf.Sprintf("descendants.id IS NULL"))
+		ctes = append(ctes, sqlf.Sprintf(
+			`WITH RECURSIVE descendants AS (
+				SELECT id, parent_team_id
```
Comment (eseliger) [inline internal/database/teams.go:61]: Maybe Name the column team_id? Feels a bit clearer to me but maybe that’s just me.

File: cmd/frontend/graphqlbackend/teams.go [hunk 2]
```diff
@@ -453,6 +453,15 @@ func (r *schemaResolver) UpdateTeam(ctx context.Context, args *UpdateTeamArgs) (
 				return errors.Wrap(err, "cannot find parent team")
 			}
 			if parentTeam.ID != t.ParentTeamID {
+				parentOutsideOfTeamsDescendants, err := tx.Teams().ContainsTeam(ctx, parentTeam.ID, database.ListTeamsOpts{
```
Comment (eseliger) [inline cmd/frontend/graphqlbackend/teams.go:456]: Do we need other ListTeam properties? Or should we just make this team, candidate int32?

**Review by eseliger — APPROVED** [review]

File: internal/database/teams.go (hunk 1 above)
Comment (cbart) [reply internal/database/teams.go:61]: Good idea. Also it does not need `parent_team_id`

File: cmd/frontend/graphqlbackend/teams.go (hunk 2 above)
Comment (cbart) [reply cmd/frontend/graphqlbackend/teams.go:456]: I feel like the method call is easier to read if the parameter is passed through ops and has a name. Plus the same tooling will be useful to suggest teams that can be set as parents (using ExceptAncestorID)

File: internal/database/teams.go (hunk 1 above)
Comment (cbart) [reply internal/database/teams.go:61]: Dang. I did not push the commit with these changes :/

**Pull request merged**